Driver: Manage variable CPU clock

Benichou34 opened the pull request:
Add function z_clock_update() in the system clock driver.
This function is a void function that must be called if the CPU clock frequency is changed.
The function will calculates and updates the new tick units from the system core clock.
The CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME symbol must be defined.

erwango reviewed (changes requested):
Two things beforehand:
- Split the PR into 2 commits: kernel and soc adaptation
- You can't modify stm32cube for zephyr specific purpose. You need to find a way to make this adaptation out of stm32cube library

Benichou34 commented:
Hello @erwango 
- I have splitted into 2 commits.
_ I don't modifiy stm32cube anymore. Now I define HSE_VALUE as the same value of CONFIG_CLOCK_STM32_HSE_CLOCK in compilation options.

erwango reviewed (approved):
Ok, from pure STM32 porting point of view

andrewboie commented on include/drivers/system_timer.h:
```diff
@@ -88,6 +88,15 @@ extern int z_clock_device_ctrl(struct device *device, u32_t ctrl_command,
+ * changed during program execution. The function calculates and
+ * updates the new tick units from the system core clock.
+ */
+extern void z_clock_update(void);
```
in typical use-cases what code would call this API?

andyross commented:
Agree with @andrewboie, it would be easier to review this if it came along with a complete integration that handled the hardware event that is changing the clock frequency.

Note that the clock rate is visible to user code and so far we've always presented it as invariant, so I worry that we may need some kind of callback or notification API if the driver can't present a constant rate.  For example: if the clock change is always a single integer divisor, why not represent the internal clock as the "fast" underlying clock and multiply the systick values received from the hardware before handing them back from k_cycle_get_32()?   That won't work for a more complicated PLL, but... is that the use case?

erwango commented on ext/hal/st/stm32cube/CMakeLists.txt:
```diff
@@ -36,6 +36,10 @@ if(CONFIG_STM32_CORE_CM4)
   zephyr_compile_definitions( -DCORE_CM4 )
 endif()
 
+if(CONFIG_CLOCK_STM32_HSE_CLOCK)
```
@Benichou34 
I think this change is not related and can be done as a separate PR.
Will be merged faster

Benichou34 replied:
@erwango I have created a separated PR for this: [PR 19](https://github.com/zephyrproject-rtos/hal_stm32/pull/19)

erwango replied:
@Benichou34 , thanks. Then for current PR, I'm not sure where it will go, but it will ease things if you remove this commit.

Benichou34 committed:
Driver: Manage variable CPU clock

Add function z_clock_update() in the system clock driver.
This function is a void function that must be called if the CPU clock
frequency is changed. The function will calculates and updates the new
tick units from the system core clock.
The CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME symbol must be defined.

Signed-off-by: Benoit Leforestier <benoit.leforestier@gmail.com>

pizi-nordic reviewed (commented):
I don't think that change of system timer frequency will be so simple. Please consider the following:

Reduction of the system timer frequency causes division of the cycles per tick value by new/old frequency ratio. If the remainder of this division is not 0, the tick duration will be changed by the operation. If such situation happens, we should recalculate deltas in the timeout list. Note, that some threads might be updating timeout list at the moment of frequency change and might already hold timeout value calculated using outdated information in local variable, not visible to the kernel.

Also, as @andyross mentioned, this change will make k_cycle_get_32() interface useless. 

IMHO if the platform would like to change CPU clock frequency in runtime, it should use system clock source not affected by the frequency scaling.

erwango commented:
@Benichou34 do you consider updating this PR ?

erwango commented:
@Benichou34 I'm closing this PR. Please feel free to re-open it if need be.

AllemanMa commented:
Are there any updates on this topic? Is there perspective of adding the feature of changing the CPU clock at runtime in Zephyr?

erwango commented:
> Are there any updates on this topic? Is there perspective of adding the feature of changing the CPU clock at runtime in Z

I don't think there are any developments on that currently. I think a new issue would be required on that topic.
